File: GFG-topological-sort.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution
{
    void dfs(int node, bool vis[], vector<int> adj[], stack<int>& st) {
        vis[node] = true;
        for(int& neighbour: adj[node]) {
            if(vis[neighbour] == false) {
                dfs(neighbour,vis,adj,st);
            }
        }
        st.push(node);
    }
	public:
	vector<int> topoSort(int V, vector<int> adj[]) 
	{
	    bool vis[V] = {false};
        stack<int> st;
        for(int i=0; i<V; i++) {
            if(vis[i] == false) {
                this->dfs(i,vis,adj,st);
            }
        }
        vector<int> answer;
        while(!st.empty()) {
            answer.push_back(st.top());
            st.pop();
        }
        return answer;
	}
};

File: GFG_Topological_Sort.java
import java.util.*;

class GFG_Topological_Sort {

    public static void dfs(int source, Stack<Integer> st, int[] visited, ArrayList<ArrayList<Integer>> adj) {
        visited[source] = 1;
        for(int neighbour: adj.get(source)) {
            if(visited[neighbour] == 0) {
                dfs(neighbour, st, visited, adj);
            }
        }
        st.push(source);
    }

    static int[] topoSort(int V, ArrayList<ArrayList<Integer>> adj) {
        int[] visited = new int[V];
        Arrays.fill(visited, 0);
        Stack<Integer> st = new Stack<>();
        for(int node = 0; node < V; node++) {
            if(visited[node] == 0) {
                dfs(node, st, visited, adj);
            }
        }
        for(int i = 0; i < V; i++) {
            visited[i] = st.peek();
            st.pop();
        }
        return visited;
    }
}